### ISSUE TITLE
[bug #28837] using PSTR with c++ produces warnings

### ISSUE THREAD

**Issue by avrs-admin** [avrdudes/avr-libc#417]
Mon 08 Feb 2010 12:01:18 AM CET

warning: only initialized variables can be placed into program memory area

this seems to be a regressiong, because it happens with code, that compiled about a year ago (no idea which version, maybe ~2007) without these warnings.

this was discussed in the ML 2009/12: [http://lists.gnu.org/archive/html/avr-gcc-list/2009-12/msg00016.html](https://savannah.nongnu.orghttp://lists.gnu.org/archive/html/avr-gcc-list/2009-12/msg00016.html)

locally redefining...
#undef PROGMEM
#define PROGMEM **_attribute_**(( section(".progmem.data") ))

#undef PSTR
#define PSTR(s) (**_extension_**({static prog_char __c[] PROGMEM = (s); &__c[0];}))
... seems to fix it.

what's the "correct" approach to handle this?

[file #24236: pgmspace.h.diff](https://savannah.nongnu.org/bugs/download.php?file_id=24236)

This issue was migrated from https://savannah.nongnu.org/bugs/?28837

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Mon 08 Feb 2010 12:14:50 AM CET

Sorry, this is not an avr-libc bug but a GCC one.  It has
already been filed to GCC's bugzilla:

[http://gcc.gnu.org/bugzilla/show_bug.cgi?id=34734](https://savannah.nongnu.orghttp://gcc.gnu.org/bugzilla/show_bug.cgi?id=34734)

Until it's fixed, you have to just live with the warning.

**Comment by avrs-admin** [comment]
Volker Kuhlmann <vok>
Sat 29 Oct 2011 03:06:21 AM CEST

The approach of redefining PROGMEM to
#define PROGMEM **attribute**(( section(".progmem.data") ))
works, but is sensitive to storage class specification.

```
#include <avr/pgmspace.h>
#undef PROGMEM
#define PROGMEM __attribute__(( section(".progmem.data") ))
PROGMEM const char string1[] = "String 1";
PROGMEM char string2[] = "String 2";
int main(void) {
return 1;
}

```
gives
error: string2 causes a section type conflict

Program memory is read-only flash memory so not using const is a bug. Unfortunately it seems avr-libc is pretty buggy here...

**Comment by avrs-admin** [comment]
Volker Kuhlmann <vok>
Sat 29 Oct 2011 06:36:20 AM CEST

May I suggest to change the definition of PROGMEM for the buggy gcc versions as per the attached patch?

See **[bug #34680](https://savannah.nongnu.org/bugs/?34680)** for PSTR()

(**[file #24236](https://savannah.nongnu.org/support/download.php?file_id=24236)**)

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Sat 29 Oct 2011 10:23:16 AM CEST

As it's a GCC bug which has already been fixed there, and it only
produces a warning, I somewhat hesitate to add hacks in avr-libc for
that.

Would there be any point in moving all PSTR data (independent of the
GCC version) to .progmem.data?

**Comment by avrs-admin** [comment]
Volker Kuhlmann <vok>
Sat 29 Oct 2011 12:42:01 PM CEST

Thanks Jörg. Those warnings are getting quite irritating in eclipse when there are a lot of them!

My suggestion has one drawback: Specifying sections in local variables gives a warning too, so I'm not sure what to do. Backport the gcc fix :-) it looks like it could be quite a while before it becomes available by normal upgrade.

The alternative is to declare the progmem variable extern first and then define it, but that has the same local variable problem.

I don't see any point in changing the section name, looking at the linker script shows that pretty much as long as it start with .progmem it's much of a muchness. The only requirement is to separate variable that are const into a differently named section from those not const. No functional difference, it's only to avoid the linker error. Maybe there's another way, but I can't think of a way to deal with progmems declared in functions. :-(

**Comment by avrs-admin** [comment]
Georg-Johann Lay <gjlayde>
Sat 29 Oct 2011 01:44:38 PM CEST

> Backport the gcc fix :-)
> it looks like it could be quite a while before it becomes
> available by normal upgrade.


avr-gcc's [PR34734](https://savannah.nongnu.orghttp://gcc.gnu.org/PR34734) is fixed in 4.6.2 and GCC 4.6.2 is available, see [http://gcc.gnu.org/gcc-4.6](https://savannah.nongnu.orghttp://gcc.gnu.org/gcc-4.6)

Bugfixes in avr-gcc are not backported to all versions supported by GCC because of avr-gcc developer shortage. That is: Bugfixes are only backported from current development trunk (4.7 at the moment) to 4.6.

Even if there was a backport to 4.5, you would still have to get a 4.5.3 distribution or build it yourself, anyway, as it won't tunnel into your local installation by magic ;-)

**Comment by avrs-admin** [comment]
Volker Kuhlmann <vok>
Sat 29 Oct 2011 03:04:02 PM CEST

I wasn't suggesting someone else do the backport, but I was wondering about my chances of success with applying the same patch you made for 4.6.2 to 4.3.3. Recompiling the source rpm shouldn't be too much of an issue.

What is the reason people are not upgrading to 4.6? 4.4 and 4.5 are not recommended ([http://www.arduino.cc/playground/Learning/Linux](https://savannah.nongnu.orghttp://www.arduino.cc/playground/Learning/Linux)).
